post-trade.ipynb

```diff
@@ -0,0 +1,319 @@
+    "    now = datetime.datetime.now()\n",
+    "    s = int(now.timestamp())\n",
+    "    return split_timestamp(s * 1000000000)\n",
+    "\n",
+    "\n",
+    "def snapshot(connection, when, product):\n",
+    "    (date, time) = split_timestamp(when)\n",
+    "    query = {\n",
+    "        \"query\": \"snapshot\",\n",
+    "        \"type\": \"levels\",\n",
+    "        \"source\": \"cme\",\n",
+    "        \"product\": product,\n",
+    "        \"date\": date,\n",
+    "        \"time\": time,\n",
+    "        \"timestamp_mode\": \"exchange\",\n",
+    "        \"time_zone\": \"UTC\",\n",
+    "    }\n",
+    "\n",
+    "    def format(obj):\n",
+    "        obj = copy.copy(obj)\n",
+    "        del obj[\"product_id\"]\n",
+    "        del obj[\"name\"]\n",
+    "        del obj[\"type\"]\n",
+    "        del obj[\"snapshot\"]\n",
+    "        del obj[\"buy\"]\n",
+    "        del obj[\"implied\"]\n",
+    "        del obj[\"message_number\"]\n",
+    "        del obj[\"transaction_id\"]\n",
+    "        return hpq.format(obj)\n",
+    "\n",
+    "    arr = list(ijson.items(connection.stream(query), \"item\"))\n",
+    "\n",
+    "    def table(pred):\n",
+    "        local = list(map(format, filter(pred, arr)))\n",
+    "        if len(local) == 0:\n",
+    "            IPython.display.display(IPython.display.HTML(\"<p>Empty</p>\"))\n",
+    "        else:\n",
+    "            table = tabulate.tabulate(local, tablefmt=\"html\", headers=\"keys\")\n",
+    "            IPython.display.display(IPython.display.HTML(table))\n",
+    "\n",
+    "    IPython.display.display(IPython.display.HTML(\"<h3>Outright Bid</h3>\"))\n",
+    "    table(lambda obj: obj[\"buy\"] and not obj[\"implied\"])\n",
+    "    IPython.display.display(IPython.display.HTML(\"<h3>Outright Ask</h3>\"))\n",
+    "    table(lambda obj: not obj[\"buy\"] and not obj[\"implied\"])\n",
+    "    IPython.display.display(IPython.display.HTML(\"<h3>Implied Bid</h3>\"))\n",
+    "    table(lambda obj: obj[\"buy\"] and obj[\"implied\"])\n",
+    "    IPython.display.display(IPython.display.HTML(\"<h3>Implied Ask</h3>\"))\n",
+    "    table(lambda obj: not obj[\"buy\"] and obj[\"implied\"])\n",
+    "\n",
+    "\n",
+    "def next_top_of_book(connection, when, product):\n",
+    "    (start_date, start_time) = split_timestamp(when)\n",
+    "    (end_date, end_time) = split_now()\n",
+    "    query = {\n",
+    "        \"query\": \"topofbook\",\n",
+    "        \"source\": \"cme\",\n",
+    "        \"product\": product,\n",
+    "        \"start_date\": start_date,\n",
+    "        \"start_time\": start_time,\n",
+    "        \"end_date\": end_date,\n",
+    "        \"end_time\": end_time,\n",
+    "        \"timestamp_mode\": \"exchange\",\n",
+    "        \"time_zone\": \"UTC\",\n",
+    "    }\n",
+    "    for obj in ijson.items(connection.stream(query), \"item\"):\n",
+    "        connection.cancel()\n",
+    "        del obj[\"type\"]\n",
```

Comment:
because of how delete works, this will return an error if the key does not exist. If this is not desired, you should use `pop`, instead.